broker/src/bin/neon_broker.rs

```diff
@@ -0,0 +1,520 @@
+                            let res_resp = http1_handler(req).await;
+                            res_resp.map(|resp| resp.map(EitherBody::Right))
+                        }
+                    }
+                }))
+            }
+        }))
+        .await?;
+    Ok(())
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use neon_broker::proto::TenantTimelineId as ProtoTenantTimelineId;
+    use tokio::sync::broadcast::error::TryRecvError;
+    use utils::id::{TenantId, TimelineId};
+
+    fn msg(timeline_id: Vec<u8>) -> SafekeeperTimelineInfo {
+        SafekeeperTimelineInfo {
+            safekeeper_id: 1,
+            tenant_timeline_id: Some(ProtoTenantTimelineId {
+                tenant_id: vec![0x00; 16],
+                timeline_id,
+            }),
+            last_log_term: 0,
+            flush_lsn: 1,
+            commit_lsn: 2,
+            backup_lsn: 3,
+            remote_consistent_lsn: 4,
+            peer_horizon_lsn: 5,
+            safekeeper_connstr: "neon-1-sk-1.local:7676".to_owned(),
+            local_start_lsn: 0,
+        }
+    }
+
+    fn tli_from_u64(i: u64) -> Vec<u8> {
+        let mut timeline_id = vec![0xFF; 8];
+        timeline_id.extend_from_slice(&i.to_be_bytes());
+        timeline_id
+    }
+
+    #[tokio::test]
+    async fn test_registry() {
+        let registry = Registry {
+            shared_state: Arc::new(RwLock::new(SharedState::new(16))),
+            chan_size: 16,
+        };
+
+        // subscribe to timeline 2
+        let ttid_2 = TenantTimelineId {
+            tenant_id: TenantId::from_slice(&[0x00; 16]).unwrap(),
+            timeline_id: TimelineId::from_slice(&tli_from_u64(2)).unwrap(),
+        };
+        let sub_key_2 = SubscriptionKey::Timeline(ttid_2);
+        let mut subscriber_2 = registry.register_subscriber(sub_key_2);
+        let mut subscriber_all = registry.register_subscriber(SubscriptionKey::All);
+
+        // send two messages with different keys
+        let msg_1 = msg(tli_from_u64(1));
+        let msg_2 = msg(tli_from_u64(2));
+        let mut publisher = registry.register_publisher();
+        publisher.send_msg(&msg_1).expect("failed to send msg");
+        publisher.send_msg(&msg_2).expect("failed to send msg");
+
+        // msg with key 2 should arrive to subscriber_2
+        assert!(
+            matches!(subscriber_2.sub_rx.try_recv(), Ok(msg) if matches!(msg.tenant_timeline_id.as_ref(), Some(ttid) if parse_proto_ttid(ttid).unwrap() == ttid_2))
+        );
+        // but nothing more
+        assert!(
+            matches!(subscriber_2.sub_rx.try_recv(), Err(err) if matches!(err, TryRecvError::Empty))
+        );
+
+        // subscriber_all should receive both messages
+        assert!(matches!(subscriber_all.sub_rx.try_recv(), Ok(..)));
+        assert!(matches!(subscriber_all.sub_rx.try_recv(), Ok(..)));
+        assert!(
+            matches!(subscriber_all.sub_rx.try_recv(), Err(err) if matches!(err, TryRecvError::Empty))
+        );
```

Comment:
NIT: a bit more readable to me
```suggestion
        assert_eq!(
            subscriber_all.sub_rx.try_recv().unwrap_err(),
            TryRecvError::Empty
        );
```